move rebuild buckets from end of first iteration to beginning of second iteration

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44330 refactor intialize bucket views
* **#44326 move rebuild buckets from end of first iteration to beginning of second iteration**

Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

zhaojuanmao committed:
move rebuild buckets from end of first iteration to beginning of second iteration

Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "move rebuild buckets from end of first iteration to beginning of second iteration"


Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "move rebuild buckets from end of first iteration to beginning of second iteration"


Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "move rebuild buckets from end of first iteration to beginning of second iteration"


Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -575,6 +576,15 @@ def no_sync(self):
 
     def forward(self, *inputs, **kwargs):
+        # Calling _rebuild_buckets before forward compuation,
+        # It may allocate new buckets before deallocating old buckets
```
> It may allocate new buckets before deallocating old buckets

Why? at this point, looks like old buckets are no longer necessary. Is it possible to first delete old buckets and then allocate new buckets?

zhaojuanmao replied:
grad points to old buckets, we can not simply delete old buckets as users may not want to delete the grads

mrshenli reviewed (approved):
Would I be correct if assume the motivation is to rebuild buckets after the backward pass, to avoid increasing peak memory consumption?

zhaojuanmao commented:
yes, memory usage is low after backward pass and before forward computation, rebuildBuckets right before forward computation can save peak memory usage if we point grad to bucket_views later on

zhaojuanmao committed:
Update on "move rebuild buckets from end of first iteration to beginning of second iteration"


Part of relanding PR #41954, this refactoring is to move rebuild_buckets call from end of first iteration to beginning of second iteration

Differential Revision: [D23583017](https://our.internmc.facebook.com/intern/diff/D23583017/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23583017/)!

[ghstack-poisoned]
